snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/core/request/SnomedQueryOptimizer.java

```diff
@@ -15,169 +15,847 @@
+		hierarchyStats.filterRedundantNoFalsePositives(conceptsToExclude);
+	}	
+
+	// Compute scores, then iterate over the best candidates with an acceptable fit in decreasing order
+	private void optimizeAncestorInclusions(final BranchContext context, final SnomedHierarchyStats inclusionHierarchyStats) {
+
+		final LongKeyFloatMap ancestorScores = inclusionHierarchyStats.initAncestorScores(
+			context, 
+			conceptSet, 
+			optimizerStrategy, 
+			zoom, 
+			weight);
+
+		if (ancestorScores.isEmpty()) {
+			return;
+		}
+		
+		final Instant startTime = Instant.now(clock);
+		final Instant endTime = startTime.plus(maxRuntime);
+
+		int iteration = 1;
+		int ancestorExpressionCount = 0;
+		boolean ancestorFound = false;
+		boolean canceled = false;
+
+		while (!canceled) {
+			final String bestFitAncestor = getBestAncestor(context, ancestorScores);
+
+			if (bestFitAncestor != null) {
+				ancestorFound = true;
+				ancestorExpressionCount++;
+
+				// Convert ancestor to an inclusion expression
+				final boolean member = conceptSet.contains(bestFitAncestor);
+				final String operator = member ? "<<" : "<";
+				final QueryExpression bestFitInclusion = new QueryExpression(IDs.base62UUID(), String.format("%s %s", operator, bestFitAncestor), false);
+
+				// Add inclusion to our optimized list
+				optimizedInclusions.add(bestFitInclusion);
+
+				// "applyInclusions" is effectively inlined here because we need access to the number of concepts included
+				final Set<String> includedIds = evaluateEclToSet(context, bestFitInclusion.getQuery());
+				final Set<String> expected = Sets.intersection(includedIds, conceptSet);
+				conceptsToInclude.removeAll(expected);
+
+				final Set<String> unexpected = Sets.difference(includedIds, conceptSet);
+				conceptsToExclude.addAll(unexpected);
+
+				// Tune parameters after a successful selection
+				if (((float) conceptSet.size()) / (includedIds.size() + 1) > 1000.0f) {
+					final float newFitThreshold = fitThreshold * 0.9f;
+					if (newFitThreshold > minimumFitThreshold) {
+						fitThreshold = newFitThreshold;
+						zoom = idealClauseCount;
+						log.info("Fit threshold changed to {}, zoom is {}", fitThreshold, zoom);
+					}
+				}
+
+				// Elevate strategy after using the same one for 100 iterations (but don't promote to LOSSY from non-lossy strategies)
+				if (iteration > (optimizerStrategy.ordinal() + 1) * 100 && optimizerStrategy.ordinal() < OptimizerStrategy.values().length - 2) {
+					optimizerStrategy = OptimizerStrategy.values()[optimizerStrategy.ordinal() + 1];
+					log.info("Optimizer strategy changed to {} after {} iterations", optimizerStrategy, iteration);
+				}
+
+			} else {
+				// No acceptable ancestor in this round
+
+				if (zoom <= maxClauseCount && zoom <= conceptSet.size() && zoom != Math.round(zoom * 1.1f)) {
+					// Try adjusting zoom first
+					zoom = Math.round(zoom * 1.1f);
+					log.info("Zoom changed to {}", zoom);
```

Comment:
Awesome log messages, but it would be great to turn them to trace messages before we merge this PR to 8.x.

Reply:
I left the initial and final messages on "info" level and turned the rest to "trace".